Add API and Reconciler for Tekton Result

sm43 opened the pull request:
Closes #163 

Signed-off-by: Shivam Mukhade <smukhade@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

```release-note
Your release note here
```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->
```
User can now install Tekton Results by creating a new CR TektonResult. The installation process is documented at  https://github.com/tektoncd/operator/blob/main/docs/tekton_results.md. This is enabled only for Kubernetes Platform.
```

sm43 commented:
/cc @nikhil-thomas

savitaashture commented on config/openshift/role.yaml:
```diff
@@ -260,3 +260,21 @@ rules:
   - jobs
   verbs:
   - '*'
+- apiGroups:
+    - authentication.k8s.io
+  resources:
+    - tokenreviews
+  verbs:
+    - create
+- apiGroups:
+    - authorization.k8s.io
+  resources:
+    - subjectaccessreviews
+  verbs:
+    - create
+- apiGroups:
+    - results.tekton.dev
+  resources:
+    - '*'
+  verbs:
+    - '*'
```
I see same changes at `config/base/role.yaml` 
So if changes are at base folder then we don't require it in Openshift folder :thinking:

nikhil-thomas commented:
/approve

nikhil-thomas commented:
Looks great 👍 

A few things needs to be added:
- release not in pr description
- a doc in `docs/` mentioning that TektonResults `will not be` auton installed like other components, as the TektonResults needs 2 secrets which end user should provide.

nikhil-thomas commented:
@wlynch could you take a look. 🧑‍💻 

short recap:
- this pr adss a new clusterScoped CRD `TektonResults` 
- user will have to create an [instance of this CRD with name `result`](https://github.com/sm43/operator/blob/2998a5db64833138adea4aed6ea0a2ff195d9648/config/crs/kubernetes/result/operator_v1alpha1_result_cr.yaml) and operator will install latest available (present in operator repo) release.yaml of TektonReults
    - operator reconciler will only 1 instance (name: result) of the clusterwide CRD
- reconciler logic in short: [read TektonResults reelase.yaml](https://github.com/sm43/operator/tree/2998a5db64833138adea4aed6ea0a2ff195d9648/cmd/kubernetes/operator/kodata/tekton-result/0.2.0), change namespace to `spec.targetNamespace`, create resources on cluster
- at present, unlike other tekton components operator will not auto-install TektonResults, as TektonResults need 2 secrets to be created before install.

cc @sm43

sm43 commented:
/hold
to add unit tests for transformers

sm43 commented:
> Looks great +1
> 
> A few things needs to be added:
> 
>     * release not in pr description
> 
>     * a doc in `docs/` mentioning that TektonResults `will not be` auton installed like other components, as the TektonResults needs 2 secrets which end user should provide.

@nikhil-thomas addressed the comments.
also added a condition to check if Target Namespace in TektonResult CR is same as TektonPipeline CR, if different then installation would fail.

/hold cancel

nikhil-thomas commented:
@sm43 let us keep TektonResults CRD and contrller available only for Kubernetes now.
For OpenShift we use payload (release.yaml) created in `github.com/openshift/tektoncd-<component-name>. eg: (https://github.com/openshift/tektoncd-pipeline/blob/release-v0.22.0/openshift/release/tektoncd-pipeline-v0.22.0.yaml)

As there is no midstream project for TektonResult, lets avoid adding it to openshift builds.

/hold make this patch kubernetes build specific

nikhil-thomas commented:
/hold

savitaashture commented:
/hold cancel

sm43 commented:
/retest

sm43 commented:
/retest

wlynch commented on test/e2e/common/tektonresultdeployement_test.go:
```diff
@@ -0,0 +1,119 @@
+// +build e2e
```
nit: tektonresultdeployment_test.go

wlynch commented on docs/tekton_results.md:
````diff
@@ -0,0 +1,52 @@
+  before installing results. By default, Tekton Results expects this secret to have
+  the following properties:
+
+    - namespace: `tekton-pipelin  es`
+    - name: `tekton-results-mysql`
+    - contains the fields:
+        - `user=root`
+        - `password=<your password>`
+
+  If you are not using a particular password management strategy, the following
+  command will generate a random password for you:
+  Update namespace value in the command if Tekton Pipelines is installed in a different namespace..
+
+   ```sh
+   $ kubectl create secret generic tekton-results-mysql --namespace=tekton-pipelines --from-literal=user=root --from-literal=password=$(openssl rand -base64 20)
+   ```
+- Generate cert/key pair. 
+  Note: Feel free to use any cert management software to do this!
+
+  Tekton Results expects the cert/key pair to be stored in a [TLS Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets).
+  Update the namespace value in below export command if Tekton Pipelines is installed in a different namespace.
+   ```sh
+   # Generate new self-signed cert.
+   $ export NAMESPACE="tekton-pipelines"
+   $ openssl req -x509 \
+   -newkey rsa:4096 \
+   -keyout key.pem \
+   -out cert.pem \
+   -days 365 \
+   -nodes \
+   -subj "/CN=tekton-results-api-service.${NAMESPACE}.svc.cluster.local" \
+   -addext "subjectAltName = DNS:tekton-results-api-service.${NAMESPACE}.svc.cluster.local"
+   # Create new TLS Secret from cert.
+   $ kubectl create secret tls -n ${NAMESPACE} tekton-results-tls \
+   --cert=cert.pem \
+   --key=key.pem
+   ```
````
I'm not super familiar with operator best practices, but would it be possible to automate these steps for the user if we find that they are not present at reconcile time?

For the TLS cert, are there existing examples for how we can hook into cert-manager or something similar?

sm43 replied:
yeah that's the plan to automate this also, To start with we kept the creation of secrets a manual steps. so, user will create secrets and then TektonResult CR.
cc @nikhil-thomas

nikhil-thomas replied:
> would it be possible to automate these steps for the user if we find that they are not present at reconcile time?
that is definitely the ideal ux the operator should deliver.
we will definitely handle in in upcoming iterations. 🧑‍💻 👍 

i have created an issue to track this: Automate creation prerequisites for TektonResult install #331

wlynch commented on test/resources/tektonresults.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// AssertTektonResultCRReadyStatus verifies if the TektonResult reaches the READY status.
+func AssertTektonResultCRReadyStatus(t *testing.T, clients *utils.Clients, names utils.ResourceNames) {
```
it's a bit odd to me to have an exported func in a library that requires a `testing.T`. Can we unexport these and move them to the test file where they are used?

sm43 replied:
oh yeah make sense. It seems the for all components it is similar way. I will address this in a separate pr

sm43 commented on cmd/kubernetes/operator/kodata/tekton-result/0.2.0/release.yaml:
```diff
@@ -0,0 +1,333 @@
+        app: tekton-results-api
+        app.kubernetes.io/name: tekton-results
+        app.kubernetes.io/component: api
+        results.tekton.dev/release: "v0.1.1"
```
yep. i downloaded both but copied wrong one I guess 😅

sm43 commented:
@wlynch the release.yaml for `v0.2.0` of Results seems to have `results.tekton.dev/release: devel` and version seems to be missing. Could you please check once? 
https://github.com/tektoncd/results/releases/tag/v0.2.0

sm43 commented:
/retest

wlynch commented:
> @wlynch the release.yaml for `v0.2.0` of Results seems to have `results.tekton.dev/release: devel` and version seems to be missing. Could you please check once?
> https://github.com/tektoncd/results/releases/tag/v0.2.0

https://github.com/tektoncd/results/issues/115 to track.

Beyond the recently opened issues, nothing else is jumping out at me, so LGTM.

/lgtm

vdemeester commented:
/lgtm
